pxtblocks/fields/field_utils.ts

```diff
@@ -693,4 +693,123 @@ function inflateJRes(jres: pxt.Map<pxt.JRes | string>): pxt.Map<pxt.JRes> {
 export function clearDropDownDiv() {
     Blockly.DropDownDiv.clearContent();
     Blockly.DropDownDiv.getContentDiv().style.height = "";
-}
\ No newline at end of file
+}
+
+interface MatrixDisplayProps {
+    blocklyId: string;
+    cellClasses?: string;
+    cellWidth: number;
+    cellHeight: number;
+    cellLabel: string;
+    cellHorizontalMargin: number;
+    cellVerticalMargin: number;
+    cornerRadius: number;
+    matrixWidth: number;
+    matrixHeight: number;
+    offColor?: string;
+    padLeft?: number;
+    parentElement: SVGGElement;
+    scale?: number;
+}
+
+export function createMatrixDisplay({
+    blocklyId,
+    cellClasses,
+    cellWidth,
+    cellHeight,
+    cellLabel,
+    cellHorizontalMargin,
+    cellVerticalMargin,
+    cornerRadius = 0,
+    matrixHeight,
+    matrixWidth,
+    offColor,
+    padLeft = 0,
+    parentElement,
+    scale = 1
+}: MatrixDisplayProps) {
+    let cells: SVGRectElement[][] = [];
+    // Initialize the matrix that holds the state
+    for (let i = 0; i < matrixWidth; i++) {
+        cells.push([]);
+    }
+
+    // Create the cells of the matrix that is displayed
+    for (let y = 0; y < matrixHeight; y++) {
+        const row = pxsim.svg.child(parentElement, "g", { 'role': 'row' });
+        for (let x = 0; x < matrixWidth; x++) {
+            createCell(x, y, row);
+        }
+    }
+
+    // TODO: things to sync
+    // on-off class vs styles
+    // sizes
+    // pointer styles oneditable
+    function createCell(x: number, y: number, row: SVGElement) {
+        const tx = scale * x * (cellWidth + cellVerticalMargin) + cellVerticalMargin + padLeft;
+        const ty = scale * y * (cellHeight + cellHorizontalMargin) + cellHorizontalMargin;
```

Comment:
This is different to the original code. tx uses x and horizontal margins, not vertical. These are mixed up for tx and ty?

Reply:
Fixing this fixes the width of the showLEDs block, but breaks the melody field editor layout. This is because these two lines of code are not the same, but have been made the same in the utils function. Note the positions of tx and ty.

melody
const cellG = pxsim.svg.child(this.elt, "g", { transform: `translate(${ty} ${tx})` }) as SVGGElement;

showLEDs
const cellG = pxsim.svg.child(row, "g", { transform: `translate(${tx} ${ty})`, 'role': 'gridcell' });

This needs some careful adjustment, following the changes through.